### PR TITLE
Add silu mul kernel

### PR DESCRIPTION
**Pull request by grimoire**
- Add kernel to fuse silu and mul in MLP.
- Optimize apply_rotary kernel and rmsnorm kernel.

### CODE REVIEW

File: lmdeploy/pytorch/kernels/cuda/activation.py [hunk 1]
```diff
@@ -0,0 +1,77 @@
+# Copyright (c) OpenMMLab. All rights reserved.
+import torch
+import triton
+import triton.language as tl
+from packaging import version
+
+TRITON_VERSION = version.parse(triton.__version__)
+
+if TRITON_VERSION >= version.parse('3.0.0'):
+
```
Comment (lvhan028) [inline lmdeploy/pytorch/kernels/cuda/activation.py:10]: Do we support triton 3.0.0?

Comment (grimoire) [reply lmdeploy/pytorch/kernels/cuda/activation.py:10]: Tested on llama3

File: lmdeploy/pytorch/kernels/cuda/apply_rotary_pos_emb.py [hunk 2]
```diff
@@ -83,44 +82,44 @@ def apply_rotary_pos_emb_qk_kernel(
...
+    if head_id < BLOCK_QH:
+        q_ptr = Q + pos_offset * stride_qs
+        qe_ptr = Q_EMB + pos_offset * stride_qes
+        ql_ptrs = q_ptr[:, None] + feat_offset_l[None, :] * stride_qd
+        qh_ptrs = q_ptr[:, None] + feat_offset_h[None, :] * stride_qd
+        qel_ptrs = qe_ptr[:, None] + feat_offset_l[None, :] * stride_qed
+        qeh_ptrs = qe_ptr[:, None] + feat_offset_h[None, :] * stride_qed
+        ql_ptrs += head_id * stride_qh
+        qh_ptrs += head_id * stride_qh
+        qel_ptrs += head_id * stride_qeh
```
Comment (lvhan028) [inline lmdeploy/pytorch/kernels/cuda/apply_rotary_pos_emb.py:94]: Is it possible that `stride_qeh` is not equal to `stride_qh`?

Comment (grimoire) [reply lmdeploy/pytorch/kernels/cuda/apply_rotary_pos_emb.py:94]: q k can be slice of qkv tensor. stride can be different if output is not inplaced.

File: lmdeploy/pytorch/backends/cuda/activation.py [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+
+    def _forward_naive(self, x):
+        """forward naive."""
+        gate, up = x.chunk(2, -1)
+        return nn.functional.silu(gate, self.inplace) * up
+
+    def forward(self, x):
+        """forward."""
+
+        if x.size(-1) % 2048 != 0:
```
Comment (AllentDan) [inline lmdeploy/pytorch/backends/cuda/activation.py:23]: why only use fused kernel when `x.size(-1) % 2048 == 0`.

Comment (grimoire) [reply lmdeploy/pytorch/backends/cuda/activation.py:23]: I fixed block size in

https://github.com/InternLM/lmdeploy/blob/f6623327d8eeb85c024c93f20f7edbc793089595/lmdeploy/pytorch/kernels/cuda/activation.py#L62

I am so lazy.

**Review by AllentDan — APPROVED** [review]

File: lmdeploy/pytorch/backends/cuda/activation.py (hunk 3 above)
Comment (lvhan028) [reply lmdeploy/pytorch/backends/cuda/activation.py:23]: I didn't get it

Comment (grimoire) [reply lmdeploy/pytorch/backends/cuda/activation.py:23]: The kernel would be more complex and slow if we had to support an arbitrary input shape.
This kernel only support aligned input. Unaligned input would be computed using the default implementation.

**Pull request merged**